vms/subnet.go

```diff
@@ -0,0 +1,323 @@
+				break
+			}
+			args.log.Debug("waiting for add subnet validator (%s) tx (%s) to be accepted", nodeID, txID)
+		}
+	}
+	return nil
+}
+
+func createBlockchain(args *args, vm CustomVM) error {
+	// Create blockchain
+	genesis, err := ioutil.ReadFile(vm.Genesis)
+	if err != nil {
+		return fmt.Errorf("could not read genesis file (%s): %w", vm.Genesis, err)
+	}
+	txID, err := args.txPChainClient.CreateBlockchain(
+		args.userPass, []string{args.fundedAddress}, args.fundedAddress, args.rSubnetID,
+		vm.ID, []string{}, vm.Name, genesis,
+	)
+	if err != nil {
+		return fmt.Errorf("could not create blockchain: %w", err)
+	}
+	for {
+		select {
+		case <-args.ctx.Done():
+			return args.ctx.Err()
+		case <-time.After(loopTimeout):
+		}
+		status, _ := args.txPChainClient.GetTxStatus(txID, true)
+		if status.Status == platformvm.Committed {
+			args.log.Info("create blockchain tx (%s) accepted", txID)
+			return nil
+		}
+		args.log.Debug("waiting for create blockchain tx (%s) to be accepted", txID)
+	}
+}
+
+func finalizeBlockchain(args *args) error {
+	// Validate blockchain exists
+	blockchains, err := args.txPChainClient.GetBlockchains()
+	if err != nil {
+		return fmt.Errorf("could not query blockchains: %w", err)
+	}
+	var blockchainID ids.ID
+	for _, blockchain := range blockchains {
+		if blockchain.SubnetID == args.rSubnetID {
+			blockchainID = blockchain.ID
+			break
+		}
+	}
+	if blockchainID == (ids.ID{}) {
+		return errors.New("could not find blockchain")
+	}
+
+	if err := ensureValidating(args, blockchainID); err != nil {
+		return fmt.Errorf("error checking all nodes are validating the blockchain: %w", err)
+	}
+	if err := ensureBootstrapped(args, blockchainID); err != nil {
+		return fmt.Errorf("error checking blockchain is bootstrapped: %w", err)
+	}
+	// Print endpoints where VM is accessible
+	args.log.Info("Custom VM endpoints now accessible at:")
+	for _, n := range args.allNodes {
+		args.log.Info("%s: %s:%d/ext/bc/%s", n.GetNodeID(), n.GetURL(), n.GetAPIPort(), blockchainID.String())
+	}
+	return nil
+}
+
+func ensureValidating(args *args, blockchainID ids.ID) error {
+	statusCheckTimeout := longTimeout
+	// Ensure all nodes are validating subnet
+	for _, n := range args.allNodes {
+		nodeID := n.GetNodeID().PrefixedString(constants.NodeIDPrefix)
+		nClient := n.GetAPIClient().PChainAPI()
+		for {
+			select {
+			case <-args.ctx.Done():
+				return args.ctx.Err()
+			case <-time.After(statusCheckTimeout):
+			}
+			status, err := nClient.GetBlockchainStatus(blockchainID.String())
+			if err != nil {
+				return fmt.Errorf("error querying blockchain status: %w", err)
+			}
+			if status == platformvm.Validating {
+				// after the first acceptance, next nodes probably don't need to check that long anymore
+				statusCheckTimeout = loopTimeout
+				break
+			}
+			args.log.Debug("waiting for validating status for %s", nodeID)
+		}
+		args.log.Info("%s validating blockchain %s", nodeID, blockchainID)
+	}
```

Comment:
In all these places that we loop over all nodes, and then have an inner loop that includes listening for a context to finish, we should use an error group. The inner for loop here will be the function run by each goroutine in the error group. If one of them returns an error (due to its context being canceled, for example) then all of them should have their context canceled.